File: src/main/java/main/puzzle/Tag.java
package main.puzzle;

import main.util.Fn;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.io.Serializable;
import java.util.List;
import java.util.*;

/**
 *
 * @author Bunnyspa
 */
public class Tag implements Serializable, Comparable<Object> {

    private static final String NAME_DEFAULT = "New Tag";

    private String name;
    private Color color;

    public Tag(Color color, String name) {
        this.color = color;
        this.name = name;
    }

    public Tag() {
        Random random = new Random();
        this.color = Fn.getColor(random.nextFloat());
        this.name = NAME_DEFAULT;
    }

    public Color getColor() {
        return color;
    }

    public void setColor(Color color) {
        this.color = color;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    @NotNull
    public String toData() {
        Color c = getColor();
        return String.format("%02X%02X%02X", c.getRed(), c.getGreen(), c.getBlue()) + getName();
    }

    @NotNull
    @Override
    public String toString() {
        return toData();
    }

    @Override
    public int compareTo(@NotNull Object o) {
        Tag t = (Tag) o;
        return this.name.compareTo(t.name);
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null || this.getClass() != obj.getClass()) {
            return false;
        }
        Tag tag = (Tag) obj;
        return this.name.equals(tag.name) && this.color.getRGB() == tag.color.getRGB();

    }

    @Override
    public int hashCode() {
        int hash = 7;
        hash = 89 * hash + Objects.hashCode(this.name);
        hash = 89 * hash + Objects.hashCode(this.color.getRGB());
        return hash;
    }

    @NotNull
    public static List<Tag> getTags(@NotNull Collection<Chip> chips) {
        Set<Tag> tagSet = new HashSet<>();
        chips.forEach((c) -> tagSet.addAll(c.getTags()));
        List<Tag> tagList = new ArrayList<>(tagSet);
        Collections.sort(tagList);
        return tagList;
    }
}

File: src/main/java/main/puzzle/Tag.kt
package main.puzzle

import main.util.Fn
import java.awt.Color
import java.io.Serializable
import java.util.*

/**
 *
 * @author Bunnyspa
 */
class Tag : Serializable, Comparable<Any> {
    var name: String
    var color: Color

    constructor(color: Color, name: String) {
        this.color = color
        this.name = name
    }

    constructor() {
        val random = Random()
        color = Fn.getColor(random.nextFloat())
        name = NAME_DEFAULT
    }

    fun toData(): String {
        val c = color
        return String.format("%02X%02X%02X", c.red, c.green, c.blue) + name
    }

    override fun toString(): String {
        return toData()
    }

    override fun compareTo(o: Any): Int {
        val t = o as Tag
        return name.compareTo(t.name)
    }

    override fun equals(obj: Any?): Boolean {
        if (this === obj) {
            return true
        }
        if (obj == null || this.javaClass != obj.javaClass) {
            return false
        }
        val tag = obj as Tag
        return name == tag.name && color.rgb == tag.color.rgb
    }

    override fun hashCode(): Int {
        var hash = 7
        hash = 89 * hash + Objects.hashCode(name)
        hash = 89 * hash + Objects.hashCode(color.rgb)
        return hash
    }

    companion object {
        private const val NAME_DEFAULT = "New Tag"
        fun getTags(chips: List<Chip>): List<Tag> {
            val tagSet: MutableSet<Tag> = hashSetOf()
            chips.forEach { c -> tagSet.addAll(c.tags) }
            val list = tagSet.toMutableList()
            list.sort()
            return list
        }
    }
}